src/Types/Cbor.purs

```diff
@@ -0,0 +1,74 @@
+module Types.Cbor where
+
+import Prelude
+
+import Contract.Prelude (foldl)
+import Control.Monad.Except (Except, runExcept, throwError)
+import Control.Monad.State.Trans (StateT, evalStateT, get, put)
+import Data.Array.Partial (head)
+import Data.Either (Either)
+import Data.Int (pow)
+import Data.Int.Bits (zshr, shl, (.&.), (.|.))
+import Data.Newtype (class Newtype)
+import Partial.Unsafe (unsafePartial)
+import Types.ByteArray (ByteArray, byteArrayToIntArray, subarray, byteLength)
+import Types.CborBytes (CborBytes(CborBytes))
+
+newtype Cbor = Cbor CborBytes
+
+derive instance Newtype Cbor _
+derive newtype instance Show Cbor
+derive newtype instance Eq Cbor
+derive newtype instance Ord Cbor
+derive newtype instance Semigroup Cbor
+derive newtype instance Monoid Cbor
+
+data CborType = ByteStringType Int
+
+data CborParseError
+  = UnknownType Int
+  | UnknownBytesType Int
+  | ByteArrayTooShort Int
+
+instance Show CborParseError where
+  show = case _ of
+    UnknownType ty -> "UnknownType " <> show ty
+    UnknownBytesType minor -> "UnknownBytesType " <> show minor
+    ByteArrayTooShort extra -> "ByteArrayTooShort " <> show extra
+
+type Parser a = StateT ByteArray (Except CborParseError) a
+
+takeN :: Int -> Parser (Array Int)
+takeN n = byteArrayToIntArray <$> takeN' n
+
+takeN' :: Int -> Parser ByteArray
+takeN' n = do
+  ba <- get
+  unless (byteLength ba >= n) $ throwError $ ByteArrayTooShort
+    (n - byteLength ba)
+  put $ subarray n (byteLength ba) ba
+  pure $ subarray 0 n ba
+
+readType :: Parser Int
+readType = unsafePartial head <$> takeN 1
+
+decodeType :: Int -> Parser CborType
+decodeType rawCborType = do
+  let
+    majorType = rawCborType `zshr` 5
```

Comment:
Please add more comments, including textual description of the algorithm.

What is `majorType`, `minorType`? 

Where to find the spec?

Reply:
I've split the code up into more functions, each with more comments, and references to the spec (and my naming also matches the spec closer). I haven't added a top level textual description of the algorithm, as my personal preference is to break up code until it reads like the algorithm. If it still needs explaining, let me know